src/main/java/org/openstreetmap/atlas/geography/atlas/raw/slicing/RawAtlasSlicer.java

```diff
@@ -320,44 +875,656 @@ protected boolean isInsideWorkingBound(final AtlasEntity entity)
+        removedLine.relationIdentifiers().forEach(relationIdentifier ->
+        {
+            if (this.stagedRelations.containsKey(relationIdentifier))
+            {
+                this.stagedRelations.get(relationIdentifier).withRemovedMember(removedLine);
+            }
+        });
+    }
+
+    /**
+     * Slice a Line that qualifies as an Area by converting it to 2d geometry, calculating its
+     * slices, and creating the new sliced Lines. If it belongs to just one country or cannot be
+     * sliced, upate the ISOCountryTag appropriately instead
+     *
+     * @param line
+     *            The Line to slice as an Area
+     */
+    private void sliceArea(final Line line)
+    {
+        final Time time = Time.now();
+        final org.locationtech.jts.geom.Polygon jtsPolygon = JTS_POLYGON_CONVERTER
+                .convert(new Polygon(line.asPolyLine()));
+        final Set<org.locationtech.jts.geom.Polygon> intersectingBoundaryPolygons = getIntersectingBoundaryPolygons(
+                jtsPolygon);
+        if (intersectingBoundaryPolygons.size() == 1
+                || CountryBoundaryMap.isSameCountry(intersectingBoundaryPolygons))
+        {
+            final String countryCode = CountryBoundaryMap.getGeometryProperty(
+                    intersectingBoundaryPolygons.iterator().next(), ISOCountryTag.KEY);
+            this.stagedLines.get(line.getIdentifier()).withAddedTag(ISOCountryTag.KEY, countryCode);
+            if (!this.isInCountry.test(this.stagedLines.get(line.getIdentifier())))
+            {
+                removeLine(line);
+            }
+            return;
+        }
+
+        // only check this once we know it's likely to be sliced
+        if (jtsPolygon.isEmpty() || !jtsPolygon.isValid())
+        {
+            logger.error("Way {} had invalid geometry, removing from Atlas {}!",
+                    line.getIdentifier(), this.shardOrAtlasName);
+            removeLine(line);
+            return;
+        }
+        final SortedMap<String, Set<org.locationtech.jts.geom.Polygon>> slices = slicePolygonGeometry(
+                jtsPolygon, intersectingBoundaryPolygons);
+        if (slices.isEmpty())
+        {
+            logger.warn("Way {} had empty sliced geometry!", line.getIdentifier());
+            this.stagedLines.get(line.getIdentifier()).withTags(line.getTags())
+                    .withAddedTag(ISOCountryTag.KEY, ISOCountryTag.COUNTRY_MISSING);
+            return;
+        }
+        else if (slices.keySet().size() == 1)
+        {
+            final String countryCode = slices.keySet().iterator().next();
+            this.stagedLines.get(line.getIdentifier()).withAddedTag(ISOCountryTag.KEY, countryCode);
+            if (!this.isInCountry.test(this.stagedLines.get(line.getIdentifier())))
+            {
+                removeLine(line);
+            }
+            return;
+        }
+        else if (slices.values().size() > AbstractIdentifierFactory.IDENTIFIER_SCALE_DEFAULT)
```

Comment:
This looks like a case of very large features with many vertices and/or an intersecting country boundary with a very large number of vertices.  Maybe comment here around how often this case is expected to be triggered, how severe the error is considered to be, and whether it is indicative of bad data or just complex or highly digitized data?